File: Cdb/src/main/java/org/eso/ias/cdb/CdbReader.java
package org.eso.ias.cdb;

import org.eso.ias.cdb.pojos.*;

import java.util.Collection;
import java.util.Optional;
import java.util.Set;

/**
 * Interface to build CDB pojos from the configuration database.
 * 
 * @author acaproni
 */
public interface CdbReader {
	
	/**
	 * Get the Ias configuration from a CDB.
	 * 
	 * @return The ias configuration read from the CDB 
	 * @throws IasCdbException In case of error getting the IAS
	 */
	Optional<IasDao> getIas() throws IasCdbException;
	
	/**
	 * Get the IASIOs.
	 * 
	 * @return The IASIOs read from the CDB
	 * @throws IasCdbException In case of error getting the IASIOs
	 */
	public Optional<Set<IasioDao>> getIasios() throws IasCdbException;
	
	/**
	 * Get the IASIO with the given ID
	 * 
	 * @param id The ID of the IASIO to read the configuration
	 * @return The IASIO read from the CDB
	 * @throws IasCdbException In case of error getting the IASIO
	 */
	public Optional<IasioDao> getIasio(String id) throws IasCdbException;
	
	/**
	 * Read the supervisor configuration from the CDB. 
	 * 
	 * @param id The not null nor empty supervisor identifier
	 * @return The Supervisor read from the CDB
	 * @throws IasCdbException In case of error getting the Supervisor
	 */
	public Optional<SupervisorDao> getSupervisor(String id) throws IasCdbException;
	
	/**
	 * Read the transfer function configuration from the CDB. 
	 * 
	 * @param tf_id The not <code>null</code> nor empty transfer function identifier
	 * @return The transfer function read from the CDB
	 * @throws IasCdbException in case of error reading from the CDB
	 */
	public Optional<TransferFunctionDao> getTransferFunction(String tf_id) throws IasCdbException;

	/**
	 * Get the transfer functions.
	 *
	 * @return The transfer functions read from the CDB
	 * @throws IasCdbException In case of error getting the transfer functions
	 */
	public Optional<Set<TransferFunctionDao>> getTransferFunctions() throws IasCdbException;

	/**
	 * Read the ttemplate configuration from the CDB. 
	 * 
	 * @param template_id The not <code>null</code> nor empty identifier of the template
	 * @return The template read from the CDB
	 * @throws IasCdbException in case of error reading from the CDB
	 */
	public Optional<TemplateDao> getTemplate(String template_id) throws IasCdbException;

	/**
	 * Get the templates.
	 *
	 * @return The templates read from the CDB
	 * @throws IasCdbException In case of error getting the templates
	 */
	public Optional<Set<TemplateDao>> getTemplates() throws IasCdbException;
	
	/**
	 * Read the ASCE configuration from the CDB. 
	 * 
	 * @param id The not null nor empty ASCE identifier
	 * @return The ASCE read from the file
	 * @throws IasCdbException In case of error getting the ASCE
	 */
	public Optional<AsceDao> getAsce(String id) throws IasCdbException;
	
	/**
	 * Read the DASU configuration from the CDB. 
	 * 
	 * @param id The not null nor empty DASU identifier
	 * @return The DASU read from the file
	 * @throws IasCdbException In case of error getting the DASU
	 */
	public Optional<DasuDao> getDasu(String id) throws IasCdbException;
	
	/**
	 * Return the DASUs to deploy in the Supervisor with the given identifier
	 * 
	 * @param id The not <code>null</code> nor empty identifier of the supervisor
	 * @return A set of DASUs to deploy in the supervisor with the passed id
	 * @throws IasCdbException in case of error reading CDB or if the 
	 *                         supervisor with the give identifier does not exist
	 */
	public Set<DasuToDeployDao> getDasusToDeployInSupervisor(String id) throws IasCdbException;
	
	/**
	 * Return the ASCEs belonging to the given DASU.
	 * 
	 * @param id The not <code>null</code> nor empty identifier of the DASU
	 * @return A set of ASCEs running in the DASU with the passed id
	 * @throws IasCdbException in case of error reading CDB or if the 
	 *                         DASU with the give identifier does not exist
	 */
	public Set<AsceDao> getAscesForDasu(String id) throws IasCdbException;
	
	/**
	 * Return the IASIOs in input to the given ASCE.
	 * 
	 * @param id The not <code>null</code> nor empty identifier of the ASCE
	 * @return A set of IASIOs in input to the ASCE
	 * @throws IasCdbException in case of error reading CDB or if the 
	 *                         ASCE with the give identifier does not exist
	 */
	public Collection<IasioDao> getIasiosForAsce(String id) throws IasCdbException;

    /**
     * Get the IDs of the Supervisors.
     *
     * This method is useful to deploy the supervisors
     *
     * @return The the IDs of the supervisors read from the CDB
     * @throws IasCdbException In case of error getting the IDs of the supervisors
     */
    public Optional<Set<String>> getSupervisorIds() throws IasCdbException;

    /**
     * Get the IDs of the DASUs.
     *
     * @return The IDs of the DASUs read from the CDB
     * @throws IasCdbException In case of error getting the IDs of the DASUs
     */
    public Optional<Set<String>> getDasuIds() throws IasCdbException;

    /**
     * Get the IDs of the ASCEs.
     *
     * @return The IDs of the ASCEs read from the CDB
     * @throws IasCdbException In case of error getting the IDs of the ASCEs
     */
    public Optional<Set<String>> getAsceIds() throws IasCdbException;

	/**
	 * Return the templated IASIOs in input to the given ASCE.
	 *
	 * These inputs are the one generated by a different template than
	 * that of the ASCE
	 * (@see <A href="https://github.com/IntegratedAlarmSystem-Group/ias/issues/124">#124</A>)
	 *
	 * @param id The not <code>null</code> nor empty identifier of the ASCE
	 * @return A set of template instance of IASIOs in input to the ASCE
	 * @throws IasCdbException in case of error reading CDB or if the
	 *                         ASCE with the give identifier does not exist
	 */
	public Collection<TemplateInstanceIasioDao> getTemplateInstancesIasiosForAsce(String id) throws IasCdbException;

	/**
	 * Get the configuration of the client with the passed identifier.
	 *
	 * The configuration is passed as a string whose format depends
	 * on the client implementation.
	 *
	 * @param id The not null nor empty ID of the IAS client
	 * @return The configuration of the client
	 * @throws IasCdbException In case of error getting the configuration of the client
	 */
	public Optional<ClientConfigDao> getClientConfig(String id) throws IasCdbException;

	/**
	 * Get the configuration of the plugin with the passed identifier.
	 *
	 * The configuration of the plugin can be read from a file or from the CDB.
	 * In both cases, the configuration is returned as #PluginConfigDao.
	 * This m,ethod returns the configuration from the CDB; reading from file is
	 * not implemented here.
	 *
	 * @param id The not null nor empty ID of the IAS plugin
	 * @return The configuration of the plugin
	 * @throws IasCdbException In case of error getting the configuration of the plugin
	 */
	public Optional<PluginConfigDao> getPluginConfig(String id) throws IasCdbException;

	/**
	 * Initialize the CDB
	 */
	public void init() throws IasCdbException;
	
	/**
	 * Close the CDB and release the associated resources
	 * @throws IasCdbException
	 */
	public void shutdown() throws IasCdbException;

}

File: Cdb/src/main/python/IasCdb/CdbReader.py
import os

from IasCdb.TextFileType import FileType, TextFileType
from IasCdb.CdbTxtFiles import CdbTxtFiles
from IasCdb.Dao.IasDao import IasDao
from IasCdb.Dao.IasioDao import IasioDao

class CdbReader:
    """
    The class to read the CDB in YAML or JSON
    """
    def __init__(self, parent_folder: str) -> None:
        """
        Constructor.

        Args:
            parent_folder: the folder that contains CDB
        """
        self.parent_folder=parent_folder
        if not os.path.isdir(parent_folder):
            raise ValueError(f"Invalid parent folder {parent_folder}")
        self.cdbTxtFiles = CdbTxtFiles.from_folder(parent_folder)
        self.files_type = self.cdbTxtFiles.files_type

    def get_ias(self) -> IasDao:
        return IasDao.from_file(self.cdbTxtFiles.get_ias_file_path(),self.files_type)
	
    def get_iasios(self) -> list[IasioDao]:
        iasios_file = self.cdbTxtFiles.get_iasio_file_path()
        if not os.path.isfile(iasios_file):
            raise ValueError(f"Cannot read {iasios_file}")
        if self.files_type==FileType.YAML:
            import yaml
            with open(iasios_file) as f:
                data=yaml.safe_load(f)
        elif self.files_type==FileType.JSON:
            import json
            with open(iasios_file) as f:
                data = json.load(f)
        else :
            raise ValueError(f"Unrecognized file type {self.files_type}")
        iasios: list[IasioDao] = []
        for iasio_data in data:
            iasio = IasioDao.from_dict(iasio_data)
            iasios.append(iasio)
        return iasios

    def get_iasio(self, id: str) -> IasioDao|None:
        if not id or id.strip()=='':
            raise ValueError("Invalid null or empty IASIO ID")
        iasios = self.get_iasios()
        for iasio in iasios:
            if iasio.id==id:
                return iasio
        return None
    
    def get_supervisor(self, id: str):
        """
		Read the supervisor configuration from the CDB
		Args
            id: The not empty supervisor identifier
		"""
        raise NotImplementedError("Method not yet implemented")
	

    def get_transfer_function(self, id: str):
        """
        Read the transfer function configuration from the CDB.
        Args:
            id: The not empty transfer function identifier
        """
        raise NotImplementedError("Method not yet implemented")

    def get_transfer_functions(self):
        """
        Get the transfer functions
        """
        raise NotImplementedError("Method not yet implemented")

    def get_template(self, id: str):
        """
        Read the ttemplate configuration from the CDB.
        Args:
            id: The not empty identifier of the template
        """
        raise NotImplementedError("Method not yet implemented")

    def get_templates(self):
        """
        Get the templates.
        """
        raise NotImplementedError("Method not yet implemented")

    def get_asce(self, id: str):
        """
        Read the ASCE configuration from the CDB.

        Args:
            id: The not empty ASCE identifier
        """
        raise NotImplementedError("Method not yet implemented")

    def get_dasu(self, id: str):
        """
        Read the DASU configuration from the CDB.

        Args:
            id: The not empty DASU identifier
        """
        raise NotImplementedError("Method not yet implemented")
	
    def get_Dasus_to_deploy_in_supervisor(self, id: str):
        """
        Return the DASUs to deploy in the Supervisor with the given identifie

        Args:
            id: The not empty identifier of the supervisor
        """
        raise NotImplementedError("Method not yet implemented")

    def get_asces_for_dasu(self, id: str):
        """
        Return the ASCEs belonging to the given DASU.

        Args:
            id: The not empty identifier of the DASU
        """
        raise NotImplementedError("Method not yet implemented")

    def get_iasios_for_asce(self, id: str):
        """
        Return the IASIOs in input to the given ASCE.

        Args:
            id: The not empty identifier of the ASCE
        """
        raise NotImplementedError("Method not yet implemented")

    def getSupervisorIds(self):
       """
       Get the IDs of the Supervisors.
       """
       raise NotImplementedError("Method not yet implemented")

    def get_dasu_ids(self):
        """
        Get the IDs of the DASUs.
        """
        raise NotImplementedError("Method not yet implemented")

    def get_asce_ids(self):
        """
        Get the IDs of the ASCEs.
        """
        raise NotImplementedError("Method not yet implemented")

    def get_template_instances_iasios_for_asce(self, id: str):
        """
        Return the templated IASIOs in input to the given ASCE.

        Args:
            id: The not empty identifier of the ASCE
        """
        raise NotImplementedError("Method not yet implemented")

    def get_client_config(self, id: str):
        """
        Get the configuration of the client with the passed identifier.
        Args:
            id: The not empty ID of the IAS clien
        """
        raise NotImplementedError("Method not yet implemented")

    def get_plugin(self, id: str):
        """
        Get the configuration of the plugin with the passed identifier.

        The configuration of the plugin can be read from a file or from the CDB.
        In both cases, the configuration is returned as PluginConfigDao
        This method returns the configuration from the CDB; reading from file is
        not implemented.

        Args:
            id: he not empty ID of the IAS plugin
        """
        raise NotImplementedError("Method not yet implemented")

    def getPluginIds(self):
        """
        Get the IDs of all the plugins in the CDB
        """
        raise NotImplementedError("Method not yet implemented")

    def get_client_ids(self):
        """
        Get the IDs of all the plugins in the CDB
        """
        raise NotImplementedError("Method not yet implemented")
